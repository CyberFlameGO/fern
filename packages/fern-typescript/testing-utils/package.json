{
  "name": "@fern-typescript/testing-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:fern-api/fern.git",
    "directory": "packages/fern-typescript/testing-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "module": "src/index.ts",
  "main": "src/index.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "yarn run compile && jest --passWithNoTests",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore",
    "lint:eslint:fix": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/ir-generator": "workspace:*",
    "@fern-api/workspace-loader": "workspace:*",
    "@fern-fern/ir-model": "0.0.49",
    "@fern-typescript/commons": "workspace:*",
    "execa": "^5.1.1",
    "is-ci": "^3.0.1",
    "memfs": "^3.4.1",
    "typescript": "4.6.4"
  },
  "devDependencies": {
    "@babel/core": "^7.18.9",
    "@babel/preset-env": "^7.18.9",
    "@babel/preset-typescript": "^7.18.6",
    "@types/is-ci": "^3.0.0",
    "@types/jest": "^28.1.6",
    "@types/node": "^18.6.1",
    "depcheck": "^1.4.3",
    "eslint": "^8.20.0",
    "jest": "^28.1.3",
    "prettier": "^2.7.1",
    "ts-jest": "^28.0.7"
  }
}
