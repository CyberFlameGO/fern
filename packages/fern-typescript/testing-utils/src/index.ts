export { itFernETE } from "./itFernETE";
export { installAndCompileGeneratedProjects } from "./runEteTest";
