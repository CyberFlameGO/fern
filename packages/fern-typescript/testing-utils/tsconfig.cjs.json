{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/cjs", "rootDir": "src", "module": "CommonJS" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.cjs.json" },
    { "path": "../../ir/ir-generator/tsconfig.cjs.json" },
    { "path": "../../workspace-parser/tsconfig.cjs.json" },
    { "path": "../commons/tsconfig.cjs.json" }
  ]
}
