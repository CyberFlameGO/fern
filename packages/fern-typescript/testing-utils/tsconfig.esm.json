{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/esm", "rootDir": "src", "module": "esnext" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.esm.json" },
    { "path": "../../compiler/compiler/tsconfig.esm.json" },
    { "path": "../commons/tsconfig.esm.json" }
  ]
}
