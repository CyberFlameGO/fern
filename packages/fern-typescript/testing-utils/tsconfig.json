{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib", "rootDir": "src" },
  "include": ["./src"],
  "references": [
    { "path": "../../generation/ir/ir-generator" },
    { "path": "../../workspace-loader" },
    { "path": "../commons" }
  ]
}
