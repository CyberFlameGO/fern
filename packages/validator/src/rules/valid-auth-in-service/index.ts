export { ValidAuthInServiceRule as default } from "./valid-auth-in-service";
