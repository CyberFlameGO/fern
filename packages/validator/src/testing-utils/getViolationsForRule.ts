import { parseWorkspaceDefinition } from "@fern-api/workspace-parser";
import { visitFernYamlAst } from "@fern-api/yaml-schema";
import { createAstVisitorForRules } from "../createAstVisitorForRules";
import { Rule, RuleViolation } from "../Rule";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToDefinition: string;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToDefinition,
}: getViolationsForRule.Args): Promise<RuleViolation[]> {
    const parseResult = await parseWorkspaceDefinition({
        name: undefined,
        absolutePathToDefinition,
    });
    if (!parseResult.didSucceed) {
        throw new Error("Failed to parse workspace");
    }

    const ruleRunner = await rule.create({ workspace: parseResult.workspace });
    const violations: RuleViolation[] = [];

    for (const [relativeFilePath, contents] of Object.entries(parseResult.workspace.files)) {
        const visitor = createAstVisitorForRules({
            relativeFilePath,
            contents,
            ruleRunners: [ruleRunner],
            addViolations: (newViolations) => {
                violations.push(...newViolations);
            },
        });
        await visitFernYamlAst(contents, visitor);
    }

    return violations;
}
