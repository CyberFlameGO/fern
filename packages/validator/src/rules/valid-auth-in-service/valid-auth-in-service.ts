import { AuthSchema } from "@fern-api/yaml-schema/src/schemas";
import chalk from "chalk";
import { Rule } from "../../Rule";

export const ValidAuthInServiceRule: Rule = {
    name: "valid-auth-in-service",
    create: () => {
        return {
            httpService: ({ service, serviceName }) => {
                const uniqueAuths = new Set<AuthSchema>();
                for (const endpointId of Object.keys(service.endpoints)) {
                    const httpEndpoint = service.endpoints[endpointId];
                    const actualAuth = httpEndpoint?.["auth-override"] ?? service.auth;
                    uniqueAuths.add(actualAuth);
                }
                if (uniqueAuths.has("bearer") && uniqueAuths.has("basic")) {
                    return [
                        {
                            severity: "error",
                            message: `Service ${chalk.bold(
                                serviceName
                            )} has endpoints with both bearer and basic auth. Only one of the two can be used.`,
                        },
                    ];
                }
                return [];
            },
        };
    },
};
