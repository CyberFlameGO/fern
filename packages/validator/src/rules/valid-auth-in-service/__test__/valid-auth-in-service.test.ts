import path from "path";
import { getViolationsForRule } from "../../../testing-utils/getViolationsForRule";
import { ValidAuthInServiceRule } from "../valid-auth-in-service";

describe("valid-auth-in-service", () => {
    it("simple", async () => {
        const violations = await getViolationsForRule({
            rule: ValidAuthInServiceRule,
            absolutePathToDefinition: path.join(__dirname, "fixtures", "simple"),
        });

        expect(violations).toMatchObject([
            {
                severity: "error",
                relativeFilePath: "src/simple.yml",
                nodePath: ["services", "http", "InvalidService"],
                message: expect.stringMatching(
                    "Service InvalidService has endpoints with both bearer and basic auth. Only one of the two can be used."
                ),
            },
        ]);
    });
});
