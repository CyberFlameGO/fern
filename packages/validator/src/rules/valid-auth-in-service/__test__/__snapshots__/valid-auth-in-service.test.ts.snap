// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid-auth-in-service simple 1`] = `
Array [
  Object {
    "message": "Service [1mInvalidService[22m has endpoints with both bearer and basic auth. Only one of the two can be used.",
    "nodePath": Array [
      "services",
      "http",
      "InvalidService",
    ],
    "relativeFilePath": "src/simple.yml",
    "severity": "error",
  },
]
`;
